Add Data.IntSet.fromRange

meooow25 opened the pull request:
For #632

Benchmark on GHC 9.2.5:
```
  fromRange:       OK (0.20s)
    376  ns ±  22 ns, 4.0 KB allocated,   2 B  copied, 7.0 MB peak memory
  fromRange:small: OK (0.18s)
    10.0 ns ± 872 ps, 119 B  allocated,   0 B  copied, 7.0 MB peak memory
```

Comparing `fromDistinctAscList [1..2^12]` from #951 against `fromRange (1,2^12)`, currently it takes `24.3 μs` (~62x the time) and with fusion it would still take `4.49 μs` (~11x the time).

konsumlamm commented on containers/src/Data/IntSet/Internal.hs:
```diff
@@ -1215,6 +1216,48 @@ fromList xs
   where
     ins t x  = insert x t
 
+-- | \(O(n / W)\). Create a set from a range of integers.
+--
+-- > fromRange (l,h) == fromList [l..h]
```
```suggestion
-- | \(O(n / W)\). Create a set from a range of integers.
--
-- > fromRange (low, high) == fromList [low..high]
--
-- If @low > high@, @fromRange (low, high)@ is empty.
```

meooow25 replied:
Changed to `low`, `high` 👍 

> `If @low > high@, @fromRange (low, high)@ is empty`

I don't think this adds value though because this is already conveyed by the line above. `[low..high]` is empty if `low` > `high`.

konsumlamm replied:
> > `If @low > high@, @fromRange (low, high)@ is empty`
> 
> I don't think this adds value though because this is already conveyed by the line above. `[low..high]` is empty if `low` > `high`.

I thought it might be good to point out the special case, even if it's redundant, but I have no strong feelings on this.

meooow25 commented:
Note: I've made some changes to the algorithm. If you didn't see the previous version you can ignore this comment. I made the change because I realized that we could be forced to do $O(W)$ work even when $n$ is very small, making the complexity $O(n/W + W)$. I changed `goL` and `goR` to avoid this, so it's correctly $O(n/W)$ now. Also fixed a bug where I used `bitmapOfSuffix` instead of `bitmapOf`, the tests ran fine because on x86 the shift value is masked to 5/6 bits anyway.

treeowl commented:
Cool!
